### PR TITLE
fix: add warn log when a file failed to be ingested

### PR DESCRIPTION
**Pull request by enzo-cappa**
From https://github.com/snowflakedb/snowflake-kafka-connector/issues/549: Add a WARN level log when a file failed to be ingested according to the copy history.

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java [hunk 1]
```diff
@@ -144,7 +144,13 @@ public Map<String, InternalUtils.IngestedFileStatus> readIngestReport(final List
           if (fileStatus.containsKey(file.getPath())) {
             numOfRecords++;
 
-            fileStatus.put(file.getPath(), convertIngestStatus(file.getStatus()));
+            final InternalUtils.IngestedFileStatus ingestionStatus = convertIngestStatus(file.getStatus());
+            fileStatus.put(file.getPath(), ingestionStatus);
+            // Log errors
+            if (InternalUtils.IngestedFileStatus.FAILED.equals(ingestionStatus) || InternalUtils.IngestedFileStatus.PARTIALLY_LOADED.equals(ingestionStatus)) {
+              LOG_WARN_MSG("Failed to load file {}\n\tError: {}\n\tColumn: {}\n\tChar pos: {}",
+                      file.getPath(), file.getFirstError(), file.getFirstErrorColumnName(), file.getFirstErrorCharacterPos());
```
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java:152]: Please add null checks for the file properties as well as the IngestedFileStatus properties

**Review by sfc-gh-rcheng — CHANGES_REQUESTED** [review]
Thanks for opening this pr @enzo-cappa ! Don't forget to run the formatter

First time contributors require a CLA to merge this change. What is a good email to forward this to?

**Comment by enzo-cappa** [comment]
> Thanks for opening this pr @enzo-cappa ! Don't forget to run the formatter
> 
> First time contributors require a CLA to merge this change. What is a good email to forward this to?

enzo.cappa _at_ snyk.io would be the best email. Thanks.

**Comment by enzo-cappa** [comment]
`Don't forget to run the formatter`

What formatter do you run usually? I don't see any instructions or specific configuration on the repo.

**Comment by sfc-gh-rcheng** [comment]
@enzo-cappa Thanks for your patience on this, we've added a CLA bot to streamline the process. Please read the CLA and sign as the bot instructs.

For formatting - you will need to download the following jar and execute it. I've opened a [PR](https://github.com/snowflakedb/snowflake-kafka-connector/pull/611) to add a ./format script that will do this for you, if you'd like to just copy-paste that into your repo or wait for it to merge.
jar: https://github.com/google/google-java-format/releases/download/v1.10.0/google-java-format-1.10.0-all-deps.jar

**Comment by sfc-gh-xhuang** [comment]
Hi @enzo-cappa do you have any questions regarding the CLA and formatter?

**Comment by enzo-cappa** [comment]
Sorry for the delay. I'm looking into this now.

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java [hunk 2]
```diff
@@ -142,7 +142,16 @@ public Map<String, InternalUtils.IngestedFileStatus> readIngestReport(final List
...
-            fileStatus.put(file.getPath(), convertIngestStatus(file.getStatus()));
+            final InternalUtils.IngestedFileStatus ingestionStatus = convertIngestStatus(file.getStatus());
+            fileStatus.put(file.getPath(), ingestionStatus);
+            // Log errors
+            if (InternalUtils.IngestedFileStatus.FAILED.equals(ingestionStatus) || InternalUtils.IngestedFileStatus.PARTIALLY_LOADED.equals(ingestionStatus)) {
+              final String firstError = file.getFirstError() != null ? file.getFirstError() : "NO ERROR";
+              final String firstErrorColumnName = file.getFirstErrorColumnName() != null ? file.getFirstErrorColumnName() : "N/A";
+              final String firstErrorCharacterPos = file.getFirstErrorCharacterPos() != null ? file.getFirstErrorCharacterPos().toString() : "N/A";
+              LOG_WARN_MSG("Failed to load file {}\n\tError: {}\n\tColumn: {}\n\tChar pos: {}",
+                      file.getPath(), firstError, firstErrorColumnName, firstErrorCharacterPos);
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java:153]: this is potentially logging PII data, lets not do this. lets only print file path, failed file ingestion errors can be further looking into ingest copy history table function. https://docs.snowflake.com/en/sql-reference/functions/copy_history

**Comment by enzo-cappa** [comment]
@sfc-gh-xhuang I didn't get any message from the bot, and I see the following error log: `Error: Could not update the JSON file: Error occured when creating a pull request comment: Resource not accessible by integration`.
Is there anything I did incorrectly in this case?

**Comment by sfc-gh-rcheng** [comment]
Looking into the CLA bot issues, problem is most likely on our end

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java (hunk 2 above)
Comment (enzo-cappa) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java:153]: We've had several use cases when there is no error description in the copy history, so I was hoping to find more details in these logs.

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java [hunk 3]
```diff
@@ -142,7 +142,16 @@ public Map<String, InternalUtils.IngestedFileStatus> readIngestReport(final List
           if (fileStatus.containsKey(file.getPath())) {
             numOfRecords++;
 
-            fileStatus.put(file.getPath(), convertIngestStatus(file.getStatus()));
+            final InternalUtils.IngestedFileStatus ingestionStatus = convertIngestStatus(file.getStatus());
+            fileStatus.put(file.getPath(), ingestionStatus);
+            // Log errors
+            if (InternalUtils.IngestedFileStatus.FAILED.equals(ingestionStatus) || InternalUtils.IngestedFileStatus.PARTIALLY_LOADED.equals(ingestionStatus)) {
```
[the remark this answers is not in the document]
Comment (enzo-cappa) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java:148]: It is considered an error in https://github.com/snowflakedb/snowflake-kafka-connector/blob/f4a0994c28298a31dc6003fbc78889bc1283acb1/src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java#L905-L909
If a file is partially loaded I assume we are missing data, and the file needs to be reprocesses in order to have all the data into Snowflake.

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java (hunk 2 above)
Comment (enzo-cappa) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java]: Applied suggestion.

**Comment by sfc-gh-rcheng** [comment]
@enzo-cappa can you update your branch on master? or at least merge this commit https://github.com/snowflakedb/snowflake-kafka-connector/commit/28800c994b825cb49eb5acdf4e2935da3d3a11ae. The CLA bot was running on an expired token, so updating the file should allow it to run properly

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java]: Sounds good, my bad, I confused it with load in progress.

**Comment by enzo-cappa** [comment]
Unfortunately I keep getting errors with the CLA bot:

```
Error: graphql call to get the committers details failed: HttpError: token *** is not a legal HTTP header value
```

**Comment by sfc-gh-xhuang** [comment]
Sorry about that. It's a bit flaky, we're waiting on another fix: https://github.com/snowflakedb/snowflake-kafka-connector/pull/646

**Comment by sfc-gh-rcheng** [comment]
@enzo-cappa we've sent you the email directly with the CLA. We've also removed the CLA bot in this PR https://github.com/snowflakedb/snowflake-kafka-connector/pull/643, please update your branch with master. Thank you for your patience on this!

**Comment by enzo-cappa** [comment]
> @enzo-cappa we've sent you the email directly with the CLA. We've also removed the CLA bot in this PR #643, please update your branch with master. Thank you for your patience on this!

Done, and done.

**Comment by enzo-cappa** [comment]
By the way, this PR is still blocked by @sfc-gh-rcheng Change Request. I think I've addressed the comments already.

**Review by sfc-gh-rcheng — APPROVED** [review]

**Comment by sfc-gh-japatel** [comment]
The reason why merge gates are failing is because github doesnt have feature to share github creds on forked repos. 
Unfortunately, we will have to open the PR ourselves and run the gates. 
@sfc-gh-rcheng can you do it? 

It is going to be very straightforward to open a PR ourselves from forked repo. 

[steps https://github.com/ORIGINAL_OWNER/ORIGINAL_REPOSITORY.git
](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/configuring-a-remote-repository-for-a-fork)

**Comment by sfc-gh-rcheng** [comment]
Corresponding PR: https://github.com/snowflakedb/snowflake-kafka-connector/pull/656

cc @sfc-gh-japatel

**Comment by enzo-cappa** [comment]
Thanks for merging the code.

**Pull request closed without merging**